src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,113 @@ pub struct ConfigSpace {
 // SAFETY: `ConfigSpace` contains only PODs in `repr(C)` or `repr(transparent)`, without padding.
 unsafe impl ByteValued for ConfigSpace {}
 
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+enum AddRxBufferError {
+    /// Error while parsing new buffer: {0}
+    Parsing(#[from] IoVecError),
+    /// RX buffer is too small
+    BufferTooSmall,
+}
+
+/// A map of all the memory the guest has provided us with for performing RX
+#[derive(Debug)]
+pub(crate) struct RxBuffers {
+    // minimum size of a usable buffer for doing RX
+    min_buffer_size: u32,
```

Comment:
> But in that case nothing critically bad happens

If we make sure that the guest is VirtIO spec compliant, this will never happen. Once we parse a descriptor that is not of the correct size will we discard it. So we will avoid re-transmissions.

Reply:
> I just think this logic is an additional complexity which will never trigger if guest does everything correct

This complexity you will need to add it with mergeable buffers. You need to make sure that the guest provides buffers of at least the virtio header size. It _might_ never trigger with a correctly operating guest. But we do not assume correctly operating guests.

> , it should be expected that something is going wrong and retransmission is acceptable.

I agree. With my logic, what goes wrong is that the buffer is discarded. This way we have much better visibility on what went actually wrong. In the other way around, the event that tells us that something went wrong is external to Firecracker (TCP re-transmissions). I would like to keep it that way.

Reply:
> This complexity you will need to add it with mergeable buffers.

In mrg_rxbuf case we only need a head descriptor to point to a buffer of at least 12 bytes (or whatever the size of the header struct). And we can just assert this.

> But we do not assume correctly operating guests.

Yes, we do not assume that. But at the same time we don't need to check this assumption. Our job is to create a system that works regardless of buffer sizes. If guest does smth that causes retransmission, it is not our problem and we should not try to prevent it on our side.